### PR TITLE
WARC Upload from replay UI

### PR DESCRIPTION
**Pull request by machawk1**

### CODE REVIEW

**Comment by machawk1** [comment]
Something a little more elegant/intuitive should be done in the UI prior to merging.

Perhaps the ability to add things to the replay system via upload should be limited to admins but I can foresee a use case of trying to facilitate collaboration via uploading WARCs.

The UI elements ought to not be so prominent by default as well, since it's a side-feature. Maybe an "Upload WARC" button to toggle the expanded interface's display. There may be additional options we want to provide relating to the uploaded WARC in the future (e.g., specifying collection metadata), so hiding these now-few elements by default is likely the right way to go.

Though we anticipate drag-and-drop functionality, having this UI handy as well will make the feature more accessible.

@ibnesayeed We were supposed to talk about this in-person but I would like to know your opinion on the initial UI.

screen shot 2018-08-11 at 12 36 55 pm

**Comment by ibnesayeed** [comment]
> Perhaps the ability to add things to the replay system via upload should be limited to admins but I can foresee a use case of trying to facilitate collaboration via uploading WARCs.

This is an interesting use case, but I can see a mechanism to do this without opening it up to everyone. In the admit interface we can add a temporary token (save in the memory as it is really temporary) button. This can generate one or more random long tokens that can then be used to temporarily authorize anyone to upload using a URL like `/uploads?access-token=<random-token>`. This way, an admin can invite people to contribute and then remove the token.

**Comment by ibnesayeed** [comment]
> The UI elements ought to not be so prominent by default as well, since it's a side-feature. Maybe an "Upload WARC" button to toggle the expanded interface's display. There may be additional options we want to provide relating to the uploaded WARC in the future (e.g., specifying collection metadata), so hiding these now-few elements by default is likely the right way to go.

The WARC upload controls can go under the Options and Details section and be hidden by default. When that drawer is open then these controls would show up.

**Comment by machawk1** [comment]
@ibnesayeed I like your idea for tokenization to allow users to upload.

However, I was thinking that the WARC upload interface would be independent of the Options and Details display, as that is hidden by default (we may want to encourage WARC uploads) but mainly, we plan to move the bulk of the admin functions in options and details to a separate interface (#213).

**Comment by ibnesayeed** [comment]
The rationale why I think it should go inside of the drawer because in is an admin functionality and users of the replay should be bothered with that. Uploads are occasional and access is repeated, so surfacing non-frequent tasks up causes unnecessary clutter in the UI.

**Comment by machawk1** [comment]
@ibnesayeed Ok. We should discuss the WARC Upload interface in-person. ;)

**Comment by machawk1** [comment]
I think we also ought to consider giving the user the option to use the new CDXJ or revert to the one originally used to start the replay system. Maybe add a meta-endpoint to make available various CDXJs instead of simply the closest CDXJ resolution scheme as proposed in #473.

**Comment by ibnesayeed** [comment]
That would require scoping which falls into the collection management territory (for which we have an existing ticket) and we will need to support index directories/files instead of a single index file in the replay.

**Comment by machawk1** [comment]
@ibnesayeed WARC upload interface moved into the "drawer" per your suggestion.
screen shot 2018-08-14 at 6 02 12 pm

**Comment by machawk1** [comment]
I believe the sole comment that needs to be addressed is to relocate the indexing logic back into the indexer and out of util, to where it was moved within this PR. Anything other requirements to get this merged, @ibnesayeed ?

**Comment by ibnesayeed** [comment]
I have not tested it functionally yet, but on an architectural level that's the primary concern for now. Once the code is relocated and refactored as necessary, I will have another look and carefully follow the code flow.

File: ipwb/indexer.py [hunk 1]
```diff
@@ -170,6 +170,83 @@ def indexFileAt(warcPaths, encryptionKey=None,
         print('\n'.join(cdxjLines))
 
 
+def generateCDXJMetadata(cdxjLines=None):
+    metadata = ['!context ["http://tools.ietf.org/html/rfc7089"]']
```
Comment (ibnesayeed) [inline ipwb/indexer.py]: I remember mentioning somewhere that we need to change this context from the RFC to a separate URL (was that comment on your JCDL submission draft?).

File: ipwb/indexer.py [hunk 2]
```diff
@@ -170,6 +170,83 @@ def indexFileAt(warcPaths, encryptionKey=None,
...
+        'generator': "InterPlanetary Wayback v.{0}".format(ipwbVersion),
+        'created_at': '{0}'.format(datetime.datetime.now().isoformat())
+    }
+    metaVals = '!meta {0}'.format(json.dumps(metaVals))
+    metadata.append(metaVals)
+
+    return metadata
+
+
+def joinCDXJFiles(cdxjPath1, cdxjPath2, outputFilePath):
```
Comment (ibnesayeed) [inline ipwb/indexer.py]: Do we really need this new `joinCDXJFiles()` function? Can't we just call existing `indexFileAt()` function instead?

File: ipwb/indexer.py [hunk 3]
```diff
@@ -170,6 +170,83 @@ def indexFileAt(warcPaths, encryptionKey=None,
...
+        cdxjMetadataLines = generateCDXJMetadata(cdxjLines)
+        cdxjLines = cdxjMetadataLines + cdxjLines
+
+        cdxjLines = '\n'.join(cdxjLines)
+
+    with open(outputFilePath, 'w') as wfd:
+        wfd.write(cdxjLines)
+
+
+def isCDXJMetadataRecord(cdxjLine):
```
Comment (ibnesayeed) [inline ipwb/indexer.py]: This function can be as simple as:

```py
    return bool(cdxjLine.strip()) and cdxjLine.strip().startswith('!')
```

File: ipwb/indexer.py (hunk 2 above)
Comment (machawk1) [reply ipwb/indexer.py]: `indexFileAt` uses a WARC to generate a CDXJ, so no. `joinCDXJFiles()` takes two CDXJs as sources and merges them.

File: ipwb/indexer.py (hunk 3 above)
Comment (machawk1) [reply ipwb/indexer.py]: That refactoring can been performed after the PR. This function existed prior to this PR and is not original code. Please focus on the changes in this PR.

File: ipwb/indexer.py (hunk 1 above)
Comment (machawk1) [reply ipwb/indexer.py]: Perhaps but per below, this code is not new to this PR.

File: ipwb/indexer.py (hunk 3 above)
Comment (ibnesayeed) [reply ipwb/indexer.py]: In fact there can be an event shorter syntax:

```py
    return cdxjLine.strip()[:1] != '!'
```

File: ipwb/indexer.py (hunk 2 above)
Comment (ibnesayeed) [reply ipwb/indexer.py]: I think `joinCDXJFiles()` is redundant here, the `indexFileAt()` already performs merging when an outfile is specified and the outfile already exists. The function first reads whatever is in the outfile before indexing any new content. Then later truncates the file and writes the aggregated output to it. I remember implementing this functionality back when I added the outfile support.

Comment (machawk1) [reply ipwb/indexer.py]: `indexFileAt` expects a WARC path as a parameter. It also considers encryption and other aspects. We need a function that, given two CDXJ files, merge them. Perhaps the logic in `indexFileAt()` can reference this new function.

File: ipwb/indexer.py (hunk 3 above)
Comment (machawk1) [reply ipwb/indexer.py]: I agree but again, this function is not introduced in this PR.

File: ipwb/indexer.py (hunk 2 above)
Comment (ibnesayeed) [reply ipwb/indexer.py]: No, we do not need to create a temporary incremental CDXJ file and a temporary combine CDXJ file and then reassign the index file to be used by the replay and piling up the throw away old CDXJ file and the incremental CDXJ file. A lot of this code can away by simply reusing what was already there. I will comment in the replay file where we need to make changes.

File: ipwb/replay.py [hunk 4]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+        filename = secure_filename(file.filename)
+        warcPath = os.path.join(app.config['UPLOAD_FOLDER'], filename)
+        file.save(warcPath)
+
+        random.seed()
+
+        cdxjPath = '/tmp/' + ''.join(random.sample(
+            string.ascii_uppercase + string.digits * 6, 12)) + '.cdxj'
+        combinedcdxjPath = '/tmp/' + ''.join(random.sample(
+            string.ascii_uppercase + string.digits * 6, 12)) + '.cdxj'
```
Comment (ibnesayeed) [inline ipwb/replay.py]: Rm both `cdxjPath` and `combinedcdxjPath` and perhaps `random.seed()` will not be needed here either.

Continue...

File: ipwb/replay.py [hunk 5]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+
+        # TODO: Check if semaphore lock exists, alert user if so, wait
+        # Index file, produce new.cdxj
+        print('Indexing file from uploaded WARC at {0} to {1}'.format(
+            warcPath, cdxjPath))
+        indexer.indexFileAt(warcPath, outfile=cdxjPath)
```
Comment (ibnesayeed) [inline ipwb/replay.py]: Use `app.cdxjFilePath` in place of `cdxjPath`.

Continue...

File: ipwb/replay.py [hunk 6]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+        print('index created at {0}'.format(cdxjPath))
```
Comment (ibnesayeed) [inline ipwb/replay.py]: This message should say, `Index updated at {app.cdxjFilePath}`

Continue...

File: ipwb/replay.py [hunk 7]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+
+        # TODO: Create semaphore lock
+        # Join current.cdxj w/ new.cdxj, write to combined.cdxj
+        print('* Prior index file: ' + app.cdxjFilePath)
+        print('* Index file of new WARC: ' + cdxjPath)
+        print('* Combined index file (to-write): ' + combinedcdxjPath)
```
Comment (ibnesayeed) [inline ipwb/replay.py]: These logs can be removed.

Continue...

File: ipwb/replay.py [hunk 8]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+        indexer.joinCDXJFiles(
+            app.cdxjFilePath, cdxjPath, combinedcdxjPath)
+        print('Setting ipwb replay index variables')
```
Comment (ibnesayeed) [inline ipwb/replay.py]: This `joinCDXJFiles()` call is not needed.

Continue...

File: ipwb/replay.py [hunk 9]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+
+        # Set replay.index to path of combined.cdxj
+        ipwbUtils.setIPWBReplayIndexPath(combinedcdxjPath)
+        app.cdxjFilePath = combinedcdxjPath
```
Comment (ibnesayeed) [inline ipwb/replay.py]: These `setIPWBReplayIndexPath()` and `cdxjFilePath` lines can go away as we are not changing the state of there.

Continue...

File: ipwb/replay.py [hunk 10]
```diff
@@ -62,6 +74,60 @@ def setServerHeader(response):
...
+        app.cdxjFileContents = getIndexFileContents(combinedcdxjPath)
```
Comment (ibnesayeed) [inline ipwb/replay.py]: While we are not using `cdxjFileContents` anywhere, but keep it in sync, this should be updated using `app.cdxjFilePath` which by now has the updated/merged content.

File: ipwb/indexer.py (hunk 3 above)
Comment (ibnesayeed) [reply ipwb/indexer.py]: With the changes suggested in the replay and the removal of a big function in the indexer, the code is going to be a lot lighter, simpler to follow, easier to maintain, and functionally the same.

**Comment by machawk1** [comment]
@ibnesayeed I made the changes you suggested in ef8a0ca291990fcc8b72187ae2371b6c6c369fdc

**Comment by ibnesayeed** [comment]
I am looking at it and I find some issues, I am going to pull the code, revert you last commits and apply some changes then will push here.

**Comment by ibnesayeed** [comment]
I didn't revert your changes, just restored a few things here and there and fixed missing pieces.

**Comment by ibnesayeed** [comment]
Wait, there are a few other functions that can be reloacted for cleaner diffs.

**Comment by machawk1** [comment]
Hmm, we don't use the variable `app.cdxjFileContents()` anywhere after setting it. What was the purpose of restoring it?

**Comment by machawk1** [comment]
@ibnesayeed Go ahead and change them. If you would rather I do it, let me know.

**Comment by ibnesayeed** [comment]
> Hmm, we don't use the variable app.cdxjFileContents() anywhere after setting it. What was the purpose of restoring it?

I restored because the original comment hinted about using some cache, I thought you might have plans on keeping the in memory and not reading the file each time. Which can be good for small indexes, but not so good for big ones on a limited memory machine. There is a refactoring needed in `getIndexFileContents()` as well, some of its functionalities belong to the startup process.

**Comment by machawk1** [comment]
@ibnesayeed That was the original idea. Please refactor after getting this merged.

**Comment by machawk1** [comment]
@ibnesayeed Something that might not show up in the tests but I just tested with 51e595d is that `app.cdxjFileContents = getIndexFileContents(combinedcdxjPath)` is still relying on the generated index that was being created when a WARC is uploaded that consists of the old index + new index combined.

**Comment by machawk1** [comment]
Disregard, the instance was refactored out.

**Comment by ibnesayeed** [comment]
> Disregard, the instance was refactored out.

I checked possible places before making that change, they were all expecting boolean value. Also, the change I made is a superset, but for existing cases it should behave exactly as it was before.

**Pull request merged**